[Masonry] Second Possible Solution of Masonry Component

hbjORbj opened the pull request:
### 1. Problem

A. The items can have arbitrary heights. [Must-have]
B. The next item is pushed to the shortest column. [Must-have]
C. The items should be ordered by row. [Must-have]
D. The number of columns `cols` and the spacing between items `spacing` can be passed into the component. [Should-have]
E. Responsive number of columns (breakpoints) [Nice-to-have]
F. Server side rendering [Nice-to-have]
G. Performance [Nice-to-have]
H. Logical tab order [Should-have]
I. Support for lazy loading/virtualisation [Nice-to-have]

### 2. Solution

**Using CSS Grid + Configuring `grid-row-end`**

A+
B+
C+
D+
E+
F-
G- (All items must finish loading in order to do computation and apply grid spans to `MasonryItem`s. Currently, I hide all items until document fully loads everything. I think it is a good idea to accept a `loader` prop from user and display this loader instead of just hiding items. )
H+
I+

oliviertassinari commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,154 @@
+
+  const classes = useUtilityClasses(styleProps);
+
+  const computeHeight = () => {
+    const child = masonryItemRef.current.firstChild;
+    setStyleProps({
+      ...styleProps,
+      contentHeight: child?.getBoundingClientRect().height,
+    });
+  };
+
+  React.useEffect(() => {
+    if (documentReady) {
+      computeHeight();
+    }
+  }, [documentReady]); // eslint-disable-line
+
+  React.useEffect(() => {
+    window.addEventListener('resize', computeHeight);;
+    return () => {
+      window.removeEventListener('resize', computeHeight);
+    };
+  });
+  return (
```
We can use a `useResizeObserver` instead, see https://github.com/mui-org/material-ui/pull/23954#issuecomment-747411562

hbjORbj commented:
@oliviertassinari I addressed all your comments in https://github.com/mui-org/material-ui/pull/27169 in this commit of this PR: https://github.com/mui-org/material-ui/pull/27439/commits/d32d3d598801d0e2f41949079e2bd817b5ab9dc7

One thing worth mentioning is that since the images used in https://deploy-preview-27169--material-ui.netlify.app/components/image-list/#standard-image-list all have the same height, they aren't sufficient to showcase the masonry component. So I mixed the current image set with those used in `ImageList`.

oliviertassinari commented on package.json:
```diff
@@ -155,6 +155,7 @@
     "react-router-dom": "^5.2.0",
     "react-test-renderer": "^17.0.1",
     "remark": "^13.0.0",
+    "resize-observer": "^1.0.2",
```
Something is off with this. The page completely froze. Did you consider using the built-in API (giving up on the older version of Safari)?

hbjORbj replied:
Yes, but I get this error: `ReferenceError: ResizeObserver is not defined`. Do you have any clue?

hbjORbj replied:
I see that `resize-observer-polyfill` is in `yarn.lock` but not in `package.json`. Maybe some team member tried this package in the past. For consistency and because it is more widely used, I removed `resize-observer` and installed `resize-observer-polyfill` instead!

oliviertassinari commented on docs/src/pages/components/masonry/BasicMasonry.js:
```diff
@@ -0,0 +1,28 @@
+          <Box
+            sx={{
+              textAlign: 'center',
+              height,
```
```suggestion
              height: '100%',
```

hbjORbj replied:
I think this should still be `height` so that each `Box` has its own height.

oliviertassinari replied:
I propose 100% here to demonstrate that the parent of the Box has the correct height.

oliviertassinari commented on docs/src/pages/components/masonry/BasicMasonry.js:
```diff
@@ -0,0 +1,28 @@
+  return (
+    <Masonry columns={4} spacing={1}>
+      {heights.map((height, index) => (
+        <MasonryItem key={index}>
```
Shouldn't we apply the height here for SSR?
```suggestion
        <MasonryItem key={index} height={height}>
```

hbjORbj replied:
Yes, I think it is a good idea that we optionally receive `height` prop for users who want SSR.

hbjORbj replied:
In fact, now I think about it, if user passes a content whose height (e.g. 80px) is different from the height the user passes (e.g. 100px) to `MasonryItem`, there will be unwanted gap.

oliviertassinari reviewed (commented):
The page completely froze. First, probably because the images we are loading are likely x10 larger than we need. to, but I also suspect something else completely off with the resize-observer polyfill we are using.

hbjORbj commented:
@oliviertassinari  I just compared two versions: one using resize event listener and the other using `ResizeObserver`. I think using resize event is much faster. With `ResizeObserver`, as in the current demo, when you resize the window, the delay is too long.

oliviertassinari commented:
> I just compared two versions: one using resize event listener and the other using ResizeObserver. I think using resize event is much faster. With ResizeObserver, as in the current demo, when you resize the window, the delay is too long.

@hbjORbj I think that what's important is that if the content of MasonryItem changes, changing its height, then the masonry layout updates. the window *resize* event won't help.

mbrookes commented on docs/src/pages/components/masonry/ImageMasonry.js:
```diff
@@ -0,0 +1,108 @@
+      {itemData.map((item) => (
+        <MasonryItem key={item.img}>
+          <img
+            srcSet={`${item.img}?fit=crop&auto=format 1x,
```
I don't think we need `fit=crop` here. Perhaps we should limit the size of image being retrieved to improve the load time.

mbrookes commented on docs/src/pages/components/masonry/ImageMasonry.js:
```diff
@@ -0,0 +1,108 @@
+        <MasonryItem key={item.img}>
+          <img
+            srcSet={`${item.img}?fit=crop&auto=format 1x,
+                ${item.img}?fit=crop&auto=format&dpr=2 2x`}
```
No advantage using `srcSet` if we're only loading a single image size.

hbjORbj replied:
That makes sense. Thank you, I will address that!

mbrookes commented on docs/src/pages/components/masonry/ImageMasonry.js:
```diff
@@ -0,0 +1,108 @@
+    title: 'Sea star',
+  },
+  {
+    img: 'https://images.pexels.com/photos/1761279/pexels-photo-1761279.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500',
```
Perhaps we should use a single source for all the images?

hbjORbj replied:
Okay, I will keep those from images.unsplash.com!

mbrookes commented on docs/src/pages/components/masonry/ImageMasonry.js:
```diff
@@ -0,0 +1,108 @@
+  return (
+    <Masonry
+      columns={{ xs: 3, sm: 5, md: 7 }}
+      spacing={{ xs: 1, sm: 2, md: 3 }}
```
I'n not sure spacing breakpoints are useful here, but if they need to be demonstrated, perhaps limit to 1 & 1.5?

hbjORbj commented:
@oliviertassinari I allowed `height` prop so that user can enjoy SSR as well and added a demo of this. I refactored `ResizeObserver` a bit and optimised images in demos for less loading time. Please let me know if image optimisation can be improved because I am not so sure if I have done the right things.

oliviertassinari commented on packages/material-ui-lab/src/MasonryItem/index.js:
```diff
@@ -0,0 +1,3 @@
+export { default } from './MasonryItem';
+export * from './masonryItemClasses';
```
```suggestion

export * from './masonryItemClasses';
```

oliviertassinari commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,166 @@
+      };
+    }
+    return true;
+  }, [masonryItemRef]);
```
You should get an eslint warning here AFAIK. Refs are stable

```suggestion
  }, []);
```

oliviertassinari commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,166 @@
+export const style = ({ styleProps, theme }) => {
+  let styles = {
+    width: '100%',
+    [`& > *`]: {
```
Simpler:

```suggestion
    '& > *': {
```

oliviertassinari commented on packages/material-ui-lab/src/Masonry/index.js:
```diff
@@ -0,0 +1,3 @@
+export { default } from './Masonry';
+export * from './masonryClasses';
```
```suggestion

export * from './masonryClasses';
```

oliviertassinari commented on package.json:
```diff
@@ -157,6 +157,7 @@
     "react-router-dom": "^5.2.0",
     "react-test-renderer": "^17.0.1",
     "remark": "^13.0.0",
+    "resize-observer-polyfill": "^1.5.1",
```
I'm continuing the discussion here

```suggestion

```

<img width="794" alt="Capture d’écran 2021-07-28 à 17 15 58" src="https://user-images.githubusercontent.com/3165635/127348415-ddf5ce0c-2963-428e-8872-8fca1d3b1eb5.png">

Where do you get this reference error?

hbjORbj replied:
If I try to use the built-in ResizeObserver without importing that polyfill package and access `http://localhost:3000/components/masonry/`, I get the following error: 
![Screenshot 2021-07-28 at 16 29 33](https://user-images.githubusercontent.com/32841130/127350986-56c691b8-3599-4571-9508-fe9221697ee5.png)

mnajdova replied:
Seeing https://caniuse.com/?search=ResizeObserver it should work :\ I think the `yarn.lock` is not pushed.

mnajdova replied:
Let's add it in the `packages/material-ui-lab/package.json`

oliviertassinari replied:
@hbjORb The error's stack trace shows that it's coming from the server. The root cause is that you can't use it server-side, there is no layout.

hbjORbj replied:
That totally makes sense. I deleted the polyfill package and we now use the built-in ResizeObserver.

oliviertassinari commented on docs/src/pages/components/masonry/ImageMasonry.tsx:
```diff
@@ -0,0 +1,96 @@
+      {itemData.map((item) => (
+        <MasonryItem key={item.img}>
+          <img
+            srcSet={`${item.img}?w=164&h=164&auto=format 1x,
+            ${item.img}?w=164&h=164&auto=format&dpr=2 2x`}
```
The image loading logic looks a LOT better. The only change I would make is to try to remove the height constraint, it seems to do nothing, w=164 wins over (it's not squared).

```suggestion
            srcSet={`${item.img}?w=164&auto=format 1x,
            ${item.img}?w=164&auto=format&dpr=2 2x`}
```

x the other cases

oliviertassinari reviewed (commented):
The resize observer logic doesn't seem to work:

https://user-images.githubusercontent.com/3165635/127349827-cbe073dd-a78c-498c-b627-91da80bea668.mp4

Otherwise, it seems that we should focus on the documentation next, and I guess we are done?

mbrookes commented on docs/src/pages/components/masonry/ImageMasonry.js:
```diff
@@ -0,0 +1,96 @@
+export default function ImageMasonry() {
+  return (
+    <Masonry
+      columns={{ xs: 3, sm: 5 }}
```
Having played with it locally, I wonder if `{ xs: 3, md: 4 }` might be the best combination?

mbrookes commented on docs/src/pages/components/masonry/ImageMasonry.js:
```diff
@@ -0,0 +1,96 @@
+      {itemData.map((item) => (
+        <MasonryItem key={item.img}>
+          <img
+            srcSet={`${item.img}?w=164&h=164&auto=format 1x,
```
With the above columns, I believe the max image width is 206px (please double check). No need to specify height.

hbjORbj commented:
@oliviertassinari Here is a fix for correctly using ResizeObserver! https://github.com/mui-org/material-ui/pull/27439/commits/7a4ddbdb834d668beb3b68ae746acf08e04ddb49


https://user-images.githubusercontent.com/32841130/127404518-6a835861-31c0-4004-8ede-09923d90a879.mov

mnajdova commented on docs/src/pages/components/masonry/masonry.md:
```diff
@@ -0,0 +1,29 @@
+
+{{"demo": "pages/components/masonry/DiffColSizeMasonry.js", "bg": true}}
+
+> In order to use server-side rendering, you can pass the height of the content of `<MasonryItem />` as shown in the following demo:
```
```suggestion
> In order to use server-side rendering, you should pass the height of the content of `<MasonryItem />` as shown in the following demo:
```

mnajdova commented on docs/src/pages/components/masonry/DiffColSizeMasonry.tsx:
```diff
@@ -0,0 +1,44 @@
+}
+
+const itemData = [
+  { height: 150, span: 1 },
+  { height: 30, span: 1 },
+  { height: 90, span: 2 },
+  { height: 110, span: 1 },
+  { height: 150, span: 1 },
+  { height: 150, span: 1 },
+  { height: 130, span: 2 },
+  { height: 80, span: 2 },
+  { height: 50, span: 1 },
+  { height: 90, span: 1 },
+  { height: 100, span: 2 },
+  { height: 150, span: 1 },
+  { height: 50, span: 1 },
+  { height: 50, span: 2 },
+  { height: 50, span: 1 },
```
```suggestion
  { height: 150 },
  { height: 30 },
  { height: 90, span: 2 },
  { height: 110 },
  { height: 150 },
  { height: 150 },
  { height: 130, span: 2 },
  { height: 80, span: 2 },
  { height: 50 },
  { height: 90 },
  { height: 100, span: 2 },
  { height: 150 },
  { height: 50 },
  { height: 50, span: 2 },
  { height: 50 },
```

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.d.ts:
```diff
@@ -0,0 +1,50 @@
+export interface MasonryTypeMap<P = {}, D extends React.ElementType = 'div'> {
+  props: P & {
+    /**
+     * The content of the component, normally `<MasonryItem />`s.
```
```suggestion
     * The content of the component. It's recommended to be `<MasonryItem />`s.
```

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.d.ts:
```diff
@@ -0,0 +1,50 @@
+     */
+    columns?: ResponsiveStyleValue<number | string>;
+    /**
+     * Defines the space between children.
```
```suggestion
     * Defines the space between children. It is a factor of the theme's spacing.
```

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.js:
```diff
@@ -0,0 +1,151 @@
+    };
+  };
+
+  styles = deepmerge(
+    styles,
+    handleBreakpoints({ theme }, spacingValues, spacingStyleFromPropValue),
+  );
```
```suggestion
  styles = {
    ...styles,
    ...handleBreakpoints({ theme }, spacingValues, spacingStyleFromPropValue),
  };
```

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.js:
```diff
@@ -0,0 +1,151 @@
+  const masonry = React.useMemo(() => ({ spacing }), [spacing]);
+
+  return (
+    <MasonryContext.Provider value={masonry}>
```
```suggestion
    <MasonryContext.Provider value={{ spacing }}>
```

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.js:
```diff
@@ -0,0 +1,151 @@
+  const styleProps = { ...props, spacing, columns };
+  const classes = useUtilityClasses(styleProps);
+
+  const masonry = React.useMemo(() => ({ spacing }), [spacing]);
+
```
```suggestion
```

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.test.js:
```diff
@@ -0,0 +1,149 @@
+      inheritComponent: 'div',
+      render,
+      refInstanceof: window.HTMLDivElement,
+      testComponentPropWith: 'div',
```
```suggestion
      testComponentPropWith: 'span',
```

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.test.js:
```diff
@@ -0,0 +1,149 @@
+    });
+  });
+
+  describe('props:', () => {
+    describe('prop: component', () => {
+      it('should render a div by default', () => {
+        const { container } = render(<Masonry>{children}</Masonry>);
+        expect(container.firstChild).to.have.property('nodeName', 'DIV');
+      });
+
+      it('should render a different component', () => {
+        const { container } = render(<Masonry component="span">{children}</Masonry>);
+        expect(container.firstChild).to.have.property('nodeName', 'SPAN');
+      });
+    });
+
+    describe('prop: className', () => {
+      it('should append the className to the root element', () => {
+        const { container } = render(<Masonry className="foo">{children}</Masonry>);
+        expect(container.firstChild).to.have.class(classes.root);
+        expect(container.firstChild).to.have.class('foo');
+      });
+    });
+  });
```
```suggestion
```

These are already being tested in the conformance test suite.

mnajdova commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.d.ts:
```diff
@@ -0,0 +1,50 @@
+export interface MasonryItemTypeMap<P = {}, D extends React.ElementType = 'div'> {
+  props: P & {
+    /**
+     * The content of the component, normally an `<img />` or a `<div />`.
```
```suggestion
     * The content of the component, normally an `<img />` or a `<div />`. It should be only one element.
```

mnajdova commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,170 @@
+      paddingBottom: gap - 1,
+    };
+  };
+  styles = deepmerge(styles, handleBreakpoints({ theme }, spacingValues, styleFromPropValue));
```
```suggestion
  styles = { ...styles, handleBreakpoints({ theme }, spacingValues, styleFromPropValue) };
```

mnajdova commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,170 @@
+    className,
+    component = 'div',
+    columnSpan = 1,
+    height = undefined,
```
```suggestion
```

It is the default when the prop is not specified

mnajdova commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.test.js:
```diff
@@ -0,0 +1,133 @@
+    });
+  });
+
+  describe('props:', () => {
+    describe('prop: component', () => {
+      it('should render a div by default', () => {
+        const { container } = render(<MasonryItem>{children}</MasonryItem>);
+        expect(container.firstChild).to.have.property('nodeName', 'DIV');
+      });
+
+      it('should render a different component', () => {
+        const { container } = render(<MasonryItem component="span">{children}</MasonryItem>);
+        expect(container.firstChild).to.have.property('nodeName', 'SPAN');
+      });
+    });
+
+    describe('prop: className', () => {
+      it('should append the className to the root element', () => {
+        const { container } = render(<MasonryItem className="foo">{children}</MasonryItem>);
+        expect(container.firstChild).to.have.class(classes.root);
+        expect(container.firstChild).to.have.class('foo');
+      });
+    });
+  });
```
```suggestion
```

Same as in the Masonry component, these are covered in the conformance test suite.

mnajdova reviewed (commented):
I can't spot any issues with the behavior, works great. Some comments for polishing, apart from that looks good 🎉

mnajdova commented:
@oliviertassinari how can we improve the new argos: https://www.argos-ci.com/mui-org/material-ui/builds/149? They are not really useful at the moment.

oliviertassinari commented:
> how can we improve the visual regression tests?

@mnajdova First, I would make `docs-components-masonry/ImageMasonry.png` and any demo with images ignored.

https://github.com/mui-org/material-ui/blob/75f27074b89381d25a1fab85fc2098af58e89cb0/test/regressions/index.js#L24

Second, I would add a `min-width: 300px;` (enough for the mobile view) to the pure layout demos.

oliviertassinari commented:
> Here is a fix for correctly using ResizeObserver! 7a4ddbd

@hbjORbj Awesome, now we can focus on the documentation. I would personally propose something in this order, based on the survey, the issue, and the benchmark:

```
# Masonry

## Basic masonry

basic demo

## Image masonry

image demo

Mention it's an alternative to https://deploy-preview-27439--material-ui.netlify.app/components/image-list/#masonry-image-list. pros vs. cons

## Spacing

demo


## Responsive values

demo

## Column spanning

demo

## Variable column width

demo

## Server-side rendering
```

hbjORbj commented:
> > Here is a fix for correctly using ResizeObserver! [7a4ddbd](https://github.com/mui-org/material-ui/commit/7a4ddbdb834d668beb3b68ae746acf08e04ddb49)
> 
> @hbjORbj Awesome, now we can focus on the documentation. I would personally propose something in this order, based on the survey, the issue, and the benchmark:
> 
> ```
> # Masonry
> 
> ## Basic masonry
> 
> basic demo
> 
> ## Image masonry
> 
> image demo
> 
> Mention it's an alternative to https://deploy-preview-27439--material-ui.netlify.app/components/image-list/#masonry-image-list. pros vs. cons
> 
> ## Spacing
> 
> demo
> 
> ## Responsive values
> 
> demo
> 
> ## Column spanning
> 
> demo
> 
> ## Variable column widths
> 
> demo
> 
> ## Server-side rendering
> 
> demo
> ```

@oliviertassinari Thank you for the structure. By "variable column widths", do you mean variable column numbers?

mnajdova commented:
> Thank you for the structure. By "variable column widths", do you mean variable column numbers?

I believe it's the `columnSpan` prop :)

oliviertassinari commented:
I was thinking of:

<img width="542" alt="Capture d’écran 2021-07-29 à 16 43 24" src="https://user-images.githubusercontent.com/3165635/127512626-ad2c39bf-e63f-4b22-8e8a-8126ebdde45b.png">

eps1lon commented on docs/src/pages/components/masonry/masonry.md:
```diff
@@ -0,0 +1,29 @@
+
+{{"demo": "pages/components/masonry/ImageMasonry.js", "bg": true}}
+
+> `<MasonryItem />` can receive not only `<img />` elements but also `<div />` elements as shown in the following demo:
```
Assuming it's not just `img` and `div`
```suggestion
> `<MasonryItem />` can receive not only `<img />` elements but any element such as `<div />` as shown in the following demo:
```

eps1lon replied:
The `>` is used for quotes. But these are paragraphs introducing the demo so I don't think we need to use `>` here. The "as shown in the following demo" also sounds bit redundant since all of these paragraphs are part of their specific demo (heading).

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.d.ts:
```diff
@@ -0,0 +1,50 @@
+    classes?: Partial<MasonryItemClasses>;
+    /**
+     * The height of the component in px.
+     * @default undefined
```
We don't document `undefined` as the default explicitly.
```suggestion
```

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,158 @@
+    gridColumnEnd: `span ${styleProps.columnSpan}`,
+  };
+
+  const base = Object.keys(theme.breakpoints.values).reduce((acc, breakpoint) => {
+    if (styleProps.spacing[breakpoint] != null) {
+      acc[breakpoint] = true;
+    }
+    return acc;
+  }, {});
```
I understand that there exists a couple of `.reduce` in the codebase using a similar pattern. But this pattern is needlessly complicated. We don't reduce anything here. This really should be just a `forEach`:

```suggestion
  const base = {};
  const base = Object.keys(theme.breakpoints.values).forEach((breakpoint) => {
    if (styleProps.spacing[breakpoint] != null) {
      acc[breakpoint] = true;
    }
  });
```

Saw a couple of other introductions of `reduce` that aren't actually reducing a list.

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,158 @@
+            });
+          })
+        : null; // If height is passed by user, ResizeObserver is not used.
+    if (resizeObserver && masonryItemRef?.current) {
```
```suggestion
    if (resizeObserver && masonryItemRef.current) {
```
`masonryItemRef` is always defined.

What should happen if the instance isn't resolved yet?

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,158 @@
+      };
+    }
+    return true;
+  }, []); // eslint-disable-line
```
Why? If you just want to re-run the effect when `height === undefined` changes then you can extract it into a separate value. But I'm pretty sure the effect needs to be synced with `height`.

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.test.js:
```diff
@@ -0,0 +1,113 @@
+import { createTheme } from '@material-ui/core/styles';
+import defaultTheme from '@material-ui/core/styles/defaultTheme';
+import ResizeObserver from 'resize-observer-polyfill';
+import { style } from './MasonryItem';
```
All of the tests using this should probably be visual regression tests. The unit tests look like they will fail every time we change the implementation slightly.

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.test.js:
```diff
@@ -0,0 +1,113 @@
+import { style } from './MasonryItem';
+
+// Mount ResizeObserver; otherwise, this error comes up: `[ReferenceError: ResizeObserver is not defined]`
+global.ResizeObserver = ResizeObserver;
```
Let's move this into a `before` and clean it up in a `after`. Otherwise every subsequent test will have `ResizeObserver` being defined when that's not the default.

We probably also want to add a descriptive error when `ResizeObserver` isn't defined. It's not supported by all browsers we support so an error is generally unexpected.

eps1lon reviewed (changes requested):
Great work overall. Really like the demos with the numbered tiles. Really helps understand what a masonry layout is.

The docs are missing what the difference is between `<Masonry />` and `<ImageList variant="masonry" />`

mbrookes commented:
There seem to be overlapping images at the bottom left of the image demo:

![image](https://user-images.githubusercontent.com/357702/127912487-8b09b931-121e-4236-a67b-e87d39d09f90.png)

mbrookes commented on docs/src/pages/components/masonry/ImageMasonry.tsx:
```diff
@@ -0,0 +1,92 @@
+
+export default function ImageMasonry() {
+  return (
+    <Masonry columns={3} spacing={1} sx={{ height: 300 }}>
```
```suggestion
    <Masonry columns={3} spacing={1} sx={{ height: 400 }}>
```

to be able to see all of at least one portrait image

mbrookes commented on docs/src/pages/components/masonry/ImageMasonry.tsx:
```diff
@@ -0,0 +1,92 @@
+      {itemData.map((item) => (
+        <MasonryItem key={item.img}>
+          <img
+            srcSet={`${item.img}?w=220&auto=format 1x,
```
```suggestion
            srcSet={`${item.img}?w=301&auto=format 1x,
```

mbrookes commented on docs/src/pages/components/masonry/ImageMasonry.tsx:
```diff
@@ -0,0 +1,92 @@
+        <MasonryItem key={item.img}>
+          <img
+            srcSet={`${item.img}?w=220&auto=format 1x,
+            ${item.img}?w=220&auto=format&dpr=2 2x`}
```
```suggestion
            ${item.img}?w=301&auto=format&dpr=2 2x`}
```

mbrookes commented on docs/src/pages/components/masonry/masonry.md:
```diff
@@ -0,0 +1,55 @@
+{{"demo": "pages/components/masonry/BasicMasonry.js", "bg": true}}
+
+## Image Masonry
+
```
Anything specific to images that could be mentioned in the description?

mbrookes commented on docs/src/pages/components/masonry/masonry.md:
```diff
@@ -0,0 +1,55 @@
+
+## Column spanning
+
+By passing `columnSpan` to `<MasonryItem />`, you can configure the number of columns taken up by each `<MasonryItem />`.
```
Best to be honest in the description about the obvious limitation of this approach: 
![image](https://user-images.githubusercontent.com/357702/127913711-7280f1f8-9df4-4ce9-bfaf-0a7435bd6d3b.png)

(Not sure why the right border gets cut off in Codesandbox)

hbjORbj replied:
I agree. I explained its limitation with a demo. The border cut-off is appearing on codesandbox because html's `box-sizing` is not set to `border-box` while MUI demo page has.

hbjORbj replied:
I resolved it in this commit: https://github.com/mui-org/material-ui/pull/27439/commits/ac7bf9dc9ea8dbfb60dbacf96a63a99427ff00b1

mnajdova commented:
Something is off with the image demo. This is what I see:

![image](https://user-images.githubusercontent.com/4512430/128000862-fab8a67d-4a29-4270-995d-50d172ccfaa0.png)

mnajdova commented on docs/src/pages/components/masonry/masonry.md:
```diff
@@ -0,0 +1,61 @@
+
+## Image Masonry
+
+`<Masonry />` orders its children by row. If you would like to order images by column, you can use `<ImageList variant="masonry" />`, whose more details can be found in [Masonry Image List](https://next.material-ui.com/components/image-list/#masonry-image-list).
```
```suggestion
`<Masonry />` orders its children by row. If you would like to order images by column, you can use `<ImageList variant="masonry" />`. More details on this component can be found in [Masonry Image List](https://next.material-ui.com/components/image-list/#masonry-image-list).
```

eps1lon commented on packages/material-ui-lab/src/Masonry/Masonry.js:
```diff
@@ -0,0 +1,150 @@
+    };
+  };
+
+  styles = deepmerge(styles, handleBreakpoints({ theme }, columnValues, columnStyleFromPropValue));
```
Here we `deepmerge` but earlier we just spread and it's unclear why either pattern was chose.

hbjORbj replied:
Just spreaing is enough in `MasonryItem`, but here in `Masonry` we need a `deepmerge` because we resolve responsive values twice, once for `spacing` and once for `columns`.

eps1lon replied:
I'm just referring to `Masonry` here. On line 64 we use `deepmerge` but on line 54 we just spread.

hbjORbj replied:
We spread on line 54 beacuse we do not need a `deepmerge`: we haven't resolved any responsive values previously and therefore do not have media queries.

eps1lon replied:
But this isn't apparent from the code at all nor does anything guard against future change where we might have media queries.

hbjORbj replied:
Not unnecessarily using `deepmerge` was Marjia's previous comment. Should I revert?

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,174 @@
+      resizeObserver.unobserve(item);
+      item.removeEventListener('error', handleError);
+    };
+  }, [height, spacing, columnSpan]); // eslint-disable-line
```
This is still problematic. We need some more context here what you specifically disable and why you disable it.

hbjORbj replied:
`styleProps` should be included in the dependency array for eslint to stop complaining, and this causes infinite re-rendering. I will put this in a comment next to the code.

oliviertassinari replied:
@hbjORbj I think that the point of Sebastian was to restructure the logic. the infinite loop is suspicious:

<img width="772" alt="Capture d’écran 2021-08-05 à 20 13 02" src="https://user-images.githubusercontent.com/3165635/128400325-752a55e6-0b00-4de9-b974-5c0b6d6fde05.png">

hbjORbj replied:
I see. Thank you. It's fixed here: https://github.com/mui-org/material-ui/pull/27439/commits/b1c4cbc9521d587a65093c916144f30bb323fd09

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,174 @@
+      });
+    });
+    const handleError = () => {
+      const src = masonryItemRef.current.firstChild.src;
+      masonryItemRef.current.firstChild.src = src;
```
What are these lines doing?

hbjORbj replied:
In case of images, if they fail loading, this will try to load them again. I added a timestamp to prevent this being ignored from cache: https://github.com/mui-org/material-ui/pull/27439/commits/4913ad805b559c00d19f792df3365718fa61fec7

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,174 @@
+      const src = masonryItemRef.current.firstChild.src;
+      masonryItemRef.current.firstChild.src = src;
+    };
+    const item = masonryItemRef.current?.firstChild;
+    item.addEventListener('error', handleError);
```
Would be less error prone to add it to `masonryItemRef.current` instead. Because now we also need to make sure a single host element is rendered.

eps1lon commented on packages/material-ui-lab/src/Masonry/Masonry.js:
```diff
@@ -0,0 +1,150 @@
+  const classes = useUtilityClasses(styleProps);
+
+  return (
+    <MasonryContext.Provider value={{ spacing }}>
```
The context value should always be memoized. Masonry potentially has many items and we want to leverage React performance optimization when `children` didn't change but `Masonry` re-rendered.

mnajdova commented:
> There seem to be overlapping images at the bottom left of the image demo:

@mbrookes looks like it is broken on Chrome only (and Edge). Firefox and Safari seem to work. Anyway we need to fix it before merging. 

Apart from this I can't spot anything else.

oliviertassinari commented:
I have pushed a couple of commits to improve the documentation. While I was testing the SSR output, I found a regression on how we handle the placeholder for the ads. I'm doing a new PR for this.

hbjORbj commented:
Overlapping of images that occurs with a small number of columns is fixed here: https://github.com/mui-org/material-ui/pull/27439/commits/92252de23472aa6063fe2e65793b08a538299d4a.

Source of error: images should have defined dimensions for a browser to correctly lay out and paint. It can handle a number of images but has difficulty with a large number of images without pre-defined dimensions. Chrome and Edge are less patient than Firefox or Safari and hence the overlapping occurred only in those browsers, but other browsers may have had the same issue with a larger number of images to be loaded.

Fix: Give a pre-defined height of 1px to each image passed to (non-SSR) `MasonryItem`, load the image separately using `new Image()`, obtain and set this height to `styleProps` and then finally define a resize observer so that responsiveness is still correctly handled.

mbrookes commented:
> Give a pre-defined height of 1px

Would setting it to 0px work, and prevent the "flash of 1px grid" :)

hbjORbj commented:
> Since memory is limited, UAs may clamp the possible size of the grid to be within a UA-defined limit, dropping all lines outside that limit. If a grid item is placed outside this limit, its grid area must be clamped to within this limited grid.

In fact, the source of error I mentioned in the previous comment is incorrect. Chrome limits the size of grid rows to 999 at the maximum. We use a gap of 1px, so if the entire masonry's height takes up more than 999px, images are squashed and hence we've seen images overlapping at the bottom in Chrome.

I will try to come up with a solution.

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,222 @@
+    }
+    const handleImageLoadError = () => {
+      const src = masonryItemRef.current.firstChild.src;
+      masonryItemRef.current.firstChild.src = `${src}?${new Date().getTime()}`;
```
What if the `src` already has a query string?

oliviertassinari replied:
Could we remove the whole introspection on the children? Why is ResizeObserver API not enough?

eps1lon replied:
How was this conversation resolved?

hbjORbj replied:
I deleted the handler and error event listener. I thought that's why Olivier meant by "remove the whole introspection on the children"? Did I misunderstand?

hbjORbj replied:
I just realised I didn't push my commits. My bad.

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,222 @@
+  React.useEffect(() => {
+    // do not create a resize observer in case of SSR masonry
+    if (isSSR) {
+      return true;
```
```suggestion
      return;
```

Effect create functions can only return a function or void

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,222 @@
+    }
+    // do not create a resize observer until height is set by onload() callback above
+    if (isImage(children) && styleProps.height === 1) {
+      return true;
```
```suggestion
      return;
```

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,222 @@
+  });
+
+  const classes = useUtilityClasses(styleProps);
+  const resizeObserver = React.useRef(null);
```
Looks like the ref is not used. We never actually read from `resizeObserver.current`

hbjORbj replied:
We use inside `useEffect()`. https://github.com/mui-org/material-ui/pull/27439/commits/92252de23472aa6063fe2e65793b08a538299d4a#diff-f3824fe7ebb878266a976503904c0d5681fe98068e49820cf469d979ed572e18R144

hbjORbj replied:
But, I think we can define it inside useEffect and not use ref.

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,222 @@
+    }
+    // re-style if `spacing` updates
+    if (spacing !== styleProps.spacing) {
+      setStyleProps({ ...styleProps, spacing });
```
This looks like derived state from props which should not be updated in an effect since it leads to tearing: One commit includes the old `height` and new `styleProps.height`, the next commit includes the new `height` and new `styleProps.height`. You can move this into the body of the function component. Same for the other updates.

eps1lon replied:
How was this conversation resolved?

hbjORbj replied:
I just realised I didn't push my commits. My bad.

eps1lon replied:
Moved conversation to https://github.com/hbjORbj/material-ui/pull/1

hbjORbj commented:
I spent the last few days experimenting with different possible solutions, trying to come up with the most optimal solution. I have two possible options, but I don't really think either option is optimal.

First possible option is virtualisation: rendering the visible rows based on the scroll position (yes, masonry will require a height in this option) and replacing the non-visible rows with a fake row. This allows us to save the number of rows. We will have to set `grid-row-end` to `span 0` for non-visible rows and ensure that the height of fake row matches the total height of these non-vislble rows. Otherwise, items will be moving around.

Second option is simply increasing the gap size to something greater than 1px and setting the given item's height to this particular row spanning height (number of rows spanned * gap size). This allows masonry to have consistent gap between items, but this may disturb user experience as we are changing the height of given items slightly (or not depending on which gap size we decide to use). If we decide on which gap size to use depending on the given masonry item's height, this at least will allow us to not increase/decrease heights too much.

I would like to hear more of your thoughts. @mnajdova @oliviertassinari @eps1lon @mbrookes

mbrookes commented:
Or a combination of the two - default row height of 2px, max height of 2000px, virtualization above that...

hbjORbj commented:
I had a call with @mnajdova to investigate further into how we will actually design virtualisation for masonry. I will experiment further and then post an update with design and implementation in detail.

eps1lon commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,161 @@
+    return () => {
+      resizeObserver.unobserve(item);
+    };
+  }, [isSSR, height]);
```
```suggestion
  }, [isSSR]);
```

`height` as a dependency is now obsolete.

hbjORbj replied:
Gottcha. Thank you!

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.test.js:
```diff
@@ -0,0 +1,114 @@
+import { expect } from 'chai';
+import * as React from 'react';
+import { createClientRender, describeConformanceV5 } from 'test/utils';
```
```suggestion
import { createClientRender, describeConformance } from 'test/utils';
```

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.test.js:
```diff
@@ -0,0 +1,114 @@
+describe('<Masonry />', () => {
+  const render = createClientRender();
+
+  describeConformanceV5(
```
```suggestion
  describeConformance(
```

mnajdova commented on docs/src/pages/components/masonry/masonry.md:
```diff
@@ -0,0 +1,65 @@
+
+## Server-side rendering
+
+By passing `height` to `<MasonryItem />`, you can use server-side rendering. By default, `height: 100%` will be set to the content of `<MasonryItem />`. If you change this, there can be unwanted gap between `<MasonryItem />` and the content that you pass to it.
```
```suggestion
By passing `defaultHeight` to `<MasonryItem />`, you can use server-side rendering. By default, `height: 100%` will be set to the content of `<MasonryItem />`. If you change this, there can be unwanted gap between `<MasonryItem />` and the content that you pass to it.
```

mnajdova commented on packages/material-ui-lab/src/Masonry/Masonry.js:
```diff
@@ -0,0 +1,152 @@
+  });
+
+  const { children, className, component = 'div', columns = 4, spacing = 1, ...other } = props;
+  const styleProps = { ...props, spacing, columns };
```
```suggestion
  const ownerState = { ...props, spacing, columns };
```

Let's rename it everywhere to be consistent with the other components.

mnajdova commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.js:
```diff
@@ -0,0 +1,173 @@
+import { getMasonryItemUtilityClass } from './masonryItemClasses';
+import MasonryContext from '../Masonry/MasonryContext';
+
+const MockResizeObserver = () => {
```
Let's add a comment as to why this exists.

mnajdova commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.test.js:
```diff
@@ -0,0 +1,113 @@
+import * as React from 'react';
+import { createClientRender, describeConformanceV5 } from 'test/utils';
```
```suggestion
import { createClientRender, describeConformance } from 'test/utils';
```

mnajdova commented on packages/material-ui-lab/src/MasonryItem/MasonryItem.test.js:
```diff
@@ -0,0 +1,113 @@
+describe('<MasonryItem />', () => {
+  const render = createClientRender();
+
+  describeConformanceV5(
```
```suggestion
  describeConformance(
```

mnajdova reviewed (approved):
This looks like a great first iteration of the component. We can hold off the virtualization for some time. I think it would be much more valuable to get first feedback from the community. I like that we can align the publishing of this component with the v5 release :) 

One thing I may change is the order of the component in the sidebar.

<img src="https://user-images.githubusercontent.com/4512430/130064112-c7e325c7-8bb2-4f64-98be-d1a9b6e1e845.png"  width="150"/>

I would put it before the Timeline so that they are in alphabetical order.

Good job overall @hbjORbj :)

oliviertassinari commented:
I landed by change on this pure CSS method. It looks brilliant in simplicity https://tobiasahlin.com/blog/masonry-with-css/.

hbjORbj commented:
> I landed by chance on this pure CSS method. It looks brilliant in simplicity https://tobiasahlin.com/blog/masonry-with-css/ but it has one downside:
> 
> A+
> **B-**
> C+
> D+
> E+
> F+
> G+
> H+
> I+

This is quite interesting. It has one more shortcoming that the masonry must have a fixed height, which should be taller than the tallest column. You can add more items to reproduce.

https://codepen.io/tobiasahlin/pen/OGBvqW

oliviertassinari commented:
> This is quite interesting. It has one more shortcoming that the masonry must have a fixed height, which should be taller than the tallest column. You can add more items to reproduce.

We could use the same resize observer as here and manually set the height of the container. 
For solving B, assuming that we measure the height of each item, we can apply for the right `order`, on our own, proof: https://codepen.io/oliviertassinari/pen/ExXYmyy (see item 5)

We can learn more about the other shortcoming of the approach at https://github.com/gilbitron/flexmasonry/issues.

---

If I understand the current tradeoff correctly:

- The height of the item is 1px off, half the time
- The height of the container can't go above 2,000px. If it does, it fails to render

I suspect that this tradeoff won't fly, but giving it a try to see sounds like a great next step. I think that 2px error is a challenge for small blocks where design precision is important and the 2,000px height limit won't be enough for dashboards with charts, stats, tables, but I could be wrong. Before merging, I think that we should address the following:

1. Each CSS property is unique, growing the size of the CSS output. How about we move `grid-row-end: span 68;` to inline style?
2. padding-bottom: -2px is not allowed https://validator.w3.org/nu/?doc=https%3A%2F%2Fdeploy-preview-27439--material-ui.netlify.app%2Fcomponents%2Fmasonry%2F
3. splot a warning once we reach the 2,000px limit, maybe linking a GitHub issue to collect feedback for the next step. We could also add a warning about it in the documentation.

Happy to see that the ResizeObserver is now used in a canonical way, no longer doing introspection on the child 👍

eps1lon reviewed (approved):
Very nice work 👍🏻

siriwatknp commented on docs/src/pages/components/masonry/BasicMasonry.tsx:
```diff
@@ -0,0 +1,30 @@
+/* eslint-disable @typescript-eslint/no-use-before-define */
+import * as React from 'react';
+import Box from '@material-ui/core/Box';
+import Masonry from '@material-ui/lab/Masonry';
+import MasonryItem from '@material-ui/lab/MasonryItem';
+
+export default function BasicMasonry() {
+  return (
+    <Box sx={{ width: 500, minHeight: 393 }}>
+      <Masonry columns={4} spacing={1}>
+        {heights.map((height, index) => (
+          <MasonryItem key={index}>
+            <Box
+              sx={{
+                textAlign: 'center',
+                height,
+                border: 1,
+                bgcolor: 'background.paper',
+              }}
+            >
+              {index + 1}
+            </Box>
+          </MasonryItem>
+        ))}
+      </Masonry>
+    </Box>
+  );
+}
+
+const heights = [150, 30, 90, 70, 110, 150, 130, 80, 50, 90, 100, 150, 30, 50, 80];
```
cleaner.
```suggestion
import * as React from 'react';
import Box from '@material-ui/core/Box';
import Masonry from '@material-ui/lab/Masonry';
import MasonryItem from '@material-ui/lab/MasonryItem';

const heights = [150, 30, 90, 70, 110, 150, 130, 80, 50, 90, 100, 150, 30, 50, 80];

export default function BasicMasonry() {
  return (
    <Box sx={{ width: 500, minHeight: 393 }}>
      <Masonry columns={4} spacing={1}>
        {heights.map((height, index) => (
          <MasonryItem key={index}>
            <Box
              sx={{
                textAlign: 'center',
                height,
                border: 1,
                bgcolor: 'background.paper',
              }}
            >
              {index + 1}
            </Box>
          </MasonryItem>
        ))}
      </Masonry>
    </Box>
  );
}

```

siriwatknp replied:
x the rest. cc @mnajdova

hbjORbj replied:
Thanks! For the demos where the constant takes up 1 line, I did as you suggested, but for the others that take up many lines, I kept them as they were!

hbjORbj commented:
@oliviertassinari I addressed your comments. I created an [issue](https://github.com/mui-org/material-ui/issues/27934) and linked in both in warning and in docs. 

Since media queries cannot be resolved into inline-style, inline-style is used to handle `gridRowEnd` and `paddingBottom` only if `spacing` is not a responsive value (array or object). When it's a responsive value, class style is used instead same as before.

oliviertassinari commented on docs/src/pages/components/masonry/masonry.md:
```diff
@@ -0,0 +1,69 @@
+The contents are ordered by row.
+If a row is already filled with the specified number of columns, the next item starts another row, and it is added to the shortest column.
+
+{{"component": "modules/components/ComponentLinkHeader.js", "design": true}}
```
**Off-topic** (not related to this PR directly)

I wonder if we should render the design kit links here. Same question for the Stack component.

adrianmanea replied:
I think it makes sense to have both components in the design kits - detachable components from where designers add their content. This way, the layer name "Stack" or "Masonry" respectively could be easily picked up by developers to know which API to use. Also, starting-point examples could be nice to have to understand how it works.

What do you think about this idea @danilo-leal

danilo-leal replied:
It's not clear to me how would I use that on the design tool. Figma would probably be better because of auto-layout but what about the other ones?! Is it a possibility to make a quick prototype for us to get a sense of it? Asking because I don't usually make into a component layout components such as these (stack, masonry, grid, etc).

oliviertassinari replied:
Ok, I'm removing these links, for now

mbrookes commented:
> By passing `propName` to `<Component />`, you can configure...

I don't believe we use this format for other components. Please consider standardising - either change this component to match the others, or all others to match this 😉
